[숫자 야구 step1] 김민석 미션 제출합니다.

alstjr7437 opened the pull request:
## 구현
```bash
.
├── domain
│   ├── BaseBallGameController.swift    * controller
│   ├── Judgment    * Ball, Strike 판단 구현
│   │   ├── BallJudgment.swift
│   │   ├── Judgment.swift
│   │   └── StrikeJudgment.swift
│   ├── NumberGenerator    * 상대방 숫자 생성기 구현
│   │   ├── NumberGenerator.swift 
│   │   └── RandomNumberGenerator.swift 
│   ├── Opponent.swift     * 상대방 숫자 구현
│   ├── Referee.swift     * 게임 결과 반환(Judgement 들고옴)
│   └── UserBall.swift     * 유저 숫자 구현(중복, 0, 자리수 등 에러 구현)
├── main.swift
└── view
    ├── InputView.swift    * 입력 View
    └── OutputView.swift    * 출력 View
``` 


## ❓질문
1. 5e643ed8d1d7d5694cbcd2f7ed3eb64eb97c4586 에 있는 판단을 프로토콜로 만들어서 까지 구조화를 하는게 애매해서 고민이 많았습니다...
2. c060d8d55cd59100a56f1ab643300d42cac4de08 Controller의 커밋 단위를 어떻게 나눴어야할지(하나씩 연동이 되면서 구현을 하다보니 전체로 커밋을 날려버렸습니다..)
3. c060d8d55cd59100a56f1ab643300d42cac4de08 Controller의 함수를 저렇게 나눈게 잘된건지.. 궁금합니다..! 
4. Controller에서 함수가 Game -> Round 호출  Game 끝나면 Restart 호출 형식인데 위파일과 같이 함수 순서를 두면 될지(Round가 먼저와야 하는지)

alstjr7437 committed:
docs: 구현할 기능 목록 추가

alstjr7437 committed:
chore: Test 초기 세팅 삭제

alstjr7437 committed:
feat: 상대방(컴퓨터) 랜덤한 숫자 만들기 구현

alstjr7437 committed:
feat: InputView 구현

숫자를 입력 받기
다시 시작할 여부 입력 받기

alstjr7437 committed:
feat: 유저 숫자 객체 구현

alstjr7437 committed:
feat: 판단 protocol 및 볼 class 구현완료

alstjr7437 committed:
feat: 스트라이크 class 구현

alstjr7437 committed:
feat: Nothin 구현하기 완료

alstjr7437 committed:
refactor: 유저볼 입력받는 형식 리팩토링

Int에서 [Int]으로 변환 및 함수 정돈

alstjr7437 committed:
chore: Referee -> Judgement 폴더명 수정

alstjr7437 committed:
feat: 심판 추가 구현(Ball,Strike,Nothing 최종 모음)

alstjr7437 committed:
feat: output 구현하기

현재 스코어 프린트를 만드는 과정에서 잘못 만들었음을 느끼고 다시 리팩토링 예정..

alstjr7437 committed:
chore: domain으로 파일 이동

alstjr7437 committed:
feat: 재시작 입력 함수 구현

alstjr7437 committed:
feat: error 출력 구현 및 게임 출력 변경

게임 출력을
맞추지 못했을때와
맞췄을때를 나눠서 print하도록 변경

alstjr7437 committed:
feat: 컨트롤러 구현(게임 전체 구현 완료)

alstjr7437 committed:
docs: 리드미 구현 기능 반영

alstjr7437 committed:
fix: 낫씽 판단 함수 삭제

낫씽은 ball과 strike 개수로 판단 가능하기 때문에 삭제

alstjr7437 committed:
chore: 프로젝트 설정 추가

alstjr7437 committed:
refactor: 재시작 변수명 수정

SeongHoonC commented on BaseBall/BaseBall/view/OutputView.swift:
```diff
@@ -0,0 +1,38 @@
+    }
+    
+    func nowScorePrint(result: [Int]){
+        if result[0] > 0 && result[1] > 0 {
+            print("\(result[0])볼 \(result[1])스트라이크")
+        } else if result[0] > 0 {
+            print("\(result[0])볼")
+        } else if result[1] > 0 {
+            print("\(result[1])스트라이크")
+        } else {
+            print("낫씽")
+        }
```
else 를 지양하고 return 을 사용해봅시다

alstjr7437 replied:
```swift
        if result[0] == 0 && result[1] == 0 {
            print("낫씽")
            return
        }
        
        var output = ""
        if result[0] > 0 {
            output += "\(result[0])볼 "
        }
        if result[1] > 0 {
            output += "\(result[1])스트라이크"
        }
            
        print(output)
```

이런식으로 바꿨는데 더욱 깔끔한 방법이 있을까요??
이 형식이 그나마 보기 제일 깔끔한 것 같아서!

SeongHoonC replied:
오.. 되게 알고리즘 같은데 괜찮은 방법이라고 생각합니다.

SeongHoonC replied:
위와 별개로 도메인과 OutputView 를 서로 다른 사람이 개발한다면 result[0] 이 뭔지 result[1] 이 뭔지 알 수 있을까요?
swift 면 struct 를 사용해 볼 수 있겠네요!

alstjr7437 replied:
생각해보니 다른 사람이 본다면 어떤게 스트라이크고 어떤게 볼인지 모르겠네요 그부분 참고해서
1b0a0e1ee5a31ab22549917aa53a59d4a9e5c13e 구현했습니다!

SeongHoonC commented on BaseBall/BaseBall/domain/UserBall.swift:
```diff
@@ -0,0 +1,36 @@
+import Foundation
+
+class UserBall{
+    var number : [Int] = []
```
외부에서 set 가능해 보이는데 private set 을 활용해 보시는건 어떤가요?

SeongHoonC replied:
함수가 인자를 받았을 때 단번에 중복이 없음을 알게 하려면 어떻게 해야할까요?

SeongHoonC replied:
추가로 변수명은 복수형으로 써주세요.

alstjr7437 replied:
> 함수가 인자를 받았을 때 단번에 중복이 없음을 알게 하려면 어떻게 해야할까요?

따로 함수를 구현해서 중복 처리를 받았는데 생성자에서 바로 처리하는게 나을까요??
```swift
    private func hasduplication(number: [Int]) -> Bool {
        let uniqueNumber = Set(number)
        return uniqueNumber.count == BALL_LENGTH
    }
```

일단 방법은 Set으로 변환해서 길이가 3인지 테스트 하면 바로 중복이 없는 것을 알 수 있을 것 같습니다!

alstjr7437 replied:
> 외부에서 set 가능해 보이는데 private set 을 활용해 보시는건 어떤가요?

아하 private set을 이용하는 방법이 있군요 
처음에는 let으로 상수 선언을 해줬는데 함수 내에서는 init으로 값이 변경이 가능해야하니까 var로 바꿨는데 private set으로 하면 init으로 값이 설정 가능하지만 외부에서는 변경이 안되니까 좋겠네요!!

아직 객체지향과 Get, Set에 개념이 많이 부족해서 다시 책을 빌려서 공부해야할 것 같네요..

SeongHoonC replied:
```swift
protocol Car {
  var speed: Int { get set }  // 차량 속도 속성 (읽기/쓰기)
  var name: String { get }    // 차량 이름 속성 (읽기 전용)

  func speedUp(speed: Int) -> Bool  // 속도 증가 메서드 (성공 여부 반환)
}
```

alstjr7437 replied:
3b76bb55deb889021489a4363d9aa24738bd867a 일단 유저 숫자와 유저 숫자가 둘다 있길래 프로토콜로 묶어서 다시 만들어줬습니다!
아직 접근제어자에서 부족한 부분이 너무 많이 보여서 공부를 더해야할 것 같네요
1. open, public - 모든 module 해당 entity에 접근
2. internal - default, entity가 작성된 module에서만 접근
3. fileprivate - entity가 작성된 source file에서만 접근
4. private - 특정 객체에서만 사용할 수 있도록
5. private(set) - getter는 internal / setter는 private이 됨

으로 일단 정리를 해뒀고 다음에 다시 공부를 해야할 것 같아요!

SeongHoonC commented on BaseBall/BaseBall/domain/Referee.swift:
```diff
@@ -0,0 +1,22 @@
+import Foundation
+
+class Referee{
+    let ballJudgement : BallJudgment
+    let strikeJudgement : StrikeJudgment
+    
+    init(ballJudgement: BallJudgment, strikeJudgement: StrikeJudgment) {
+        self.ballJudgement = ballJudgement
+        self.strikeJudgement = strikeJudgement
+    }
+    
+    func gameResult(userNumber: [Int], computerNumber: [Int]) -> [Int]{
+        return [ballJudgement.gameCount(userNumber: userNumber, computerNumber: computerNumber), strikeJudgement.gameCount(userNumber: userNumber, computerNumber: computerNumber)]
```
BallJudgement 와 StrikeJudgement 를 Judgement 로 추상화를 했는데 결국 사용되는 곳에서 ball 인지 strike 인지 알아야하네요. 과한 추상화가 아닌지 생각해볼 필요가 있습니다.

alstjr7437 replied:
이부분에서 따로 판단을 프로토콜로 만들고 볼과 스트라이크를 만들어서 
심판에게 넘겨줘야겠다 생각하고 구현했는데 결국 Ball과 Strike의 규칙이 변경되는 일은 없더라고요 그래서 질문에도 남긴 것 처럼 과한 추상화가 아닌가 계속 의문이 있었습니다!

SeongHoonC commented on BaseBall/BaseBall/domain/NumberGenerator/RandomNumberGenerator.swift:
```diff
@@ -0,0 +1,17 @@
+
+import Foundation
+
+class RandomNumberGenerator : NumberGenerator{
+    var num = Array(1...9)
+    
+    func getNumber() -> [Int] {
+        num.shuffle()
+        return Array(num[0...2])
+    }
+}
```
이야 굉장히 깔끔하게 랜덤을 생성하셨군요! :)

alstjr7437 replied:
감사합니다 😊

SeongHoonC commented on BaseBall/BaseBall/domain/Opponent.swift:
```diff
@@ -0,0 +1,17 @@
+
+import Foundation
+
+class Opponent{
+    var number : [Int]
+    
+    init(numberGenerator : NumberGenerator){
+        self.number = numberGenerator.getNumber()
+    }
+    
+}
```
이 객체의 책임은 무엇인가요?

alstjr7437 replied:
상대방의 숫자를 가지고 있는 책임을 가지고 있는데
이번에 따로 시간이 되면 구현해볼 문제로 친구가 맞춰야할 숫자를 처음에 입력을 해두면 재밌을 것 같아서 
추후에 상대방 숫자 로직을 위해서 만들었습니다. 

어떻게 보면 UserBall의 반대되는 객체라고 생각하고 잇습니다.

일단 현재 스텝 단계에서는 randomNumberGenerator를 받았습니다

SeongHoonC commented on BaseBall/BaseBall/domain/Judgment/Judgment.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+protocol Judgment {
+    func gameCount(userNumber: [Int], computerNumber: [Int]) -> Int
+}
```
사용자인지, 컴퓨터인지..심판의 관심사일까요?

alstjr7437 replied:
> 사용자인지, 컴퓨터인지..심판의 관심사일까요?

그럼 이부분에서 userNumber, computerNumber라는 이름보다
그냥 inputNumber1, inputNumber2 이런식이 좋겠네요!

SeongHoonC replied:
음 기준과 비교 대상도 괜찮아보이네요.

SeongHoonC commented on BaseBall/BaseBall/domain/Judgment/StrikeJudgment.swift:
```diff
@@ -0,0 +1,14 @@
+import Foundation
+
+class StrikeJudgment : Judgment{
+    func gameCount(userNumber: [Int], computerNumber: [Int]) -> Int {
```
gameCount 라는 함수명이 뭘 하는지, 반환은 무엇인지 한번에 와닿지 않네요.

alstjr7437 replied:
Strike Count라고 짓기에는 Ball에서도 가져와서 쓰기때문에 저렇게 지어진 것 같네요!!
과도한 추상화를 통해서 함수명 짓기에도 애매한 부분이 생긴 것 같아요.

SeongHoonC commented on BaseBall/BaseBall/domain/BaseBallGameController.swift:
```diff
@@ -0,0 +1,70 @@
+        if number != RESTART_NUM1 && number != RESTART_NUM2{
+            throw RestartError.IllegalArgumentException
+        } else {
+            return number == 1 ? true : false
```
```swift
return number == 1 
```
이 문법은 안되나요?

alstjr7437 replied:
처음에는 1,2로 해서 1이 아니면 전부 넘겨주면서 테스트를 하다가 
1,2가 아닐때는 에러로 넘겨주는 부분을 추가하고 아래 코드를 수정하지 않아서 이러한 코드가 생긴 것 같네요!

SeongHoonC commented on BaseBall/BaseBall/domain/BaseBallGameController.swift:
```diff
@@ -0,0 +1,70 @@
+class BaseBallGameController{
+    let inputView : InputView
+    let outputView : OutputView
+    var restartJudge = true
```
restartJudge 를 상태로 가지면서 false 가 한번 들어가면 게임을 더이상 못하는 객체가 되어버렸네요.

alstjr7437 replied:
맞네요 변수를 변경해서 쓰면 안될 것 같네요!

그래서 아래와 같이 guard를 이용해서 restart()가 false면 break 하는 형식으로 바꿔줬습니다!
```swift
    func run() {
        do {
            while true {
                try playOneGame(opponentNumbers: [1,2,3])
                guard try restart() else {
                    outputView.endGamePrint()
                    break
                }
            }
            
        } catch {
            outputView.errorInputPrint()
        }
    }
```

SeongHoonC reviewed (changes requested):
고생많았어요! 전반적으로 깔끔하지만 더 좋은 방향이 많이 보여서 코멘트 남겨봤습니다.
리뷰 반영 후 재요청해주세요~

커밋을 보고 어떤 요구사항을 만족하기 위함인지 알기 어려워요.
어떤 클래스 구현, 어떤 함수 구현 이라는 커밋 메세지는 직접 구현한 본인만 알 수 있겠죠?
전혀 모르는 사람이 봐도 어떤 기능을 구현하기 위해 작업한 커밋 내역이구나 알 수 있게 남겨보세요!

alstjr7437 committed:
chore: controller 폴더 위치 변경

alstjr7437 committed:
refactor: 점수 출력 형식 변경

alstjr7437 committed:
reafactor: 변수명 전체 수정

숫자 여러개가 들어가있는 복수형이니 number -> numbers로 수정

alstjr7437 committed:
Test: Nothing Test 삭제

alstjr7437 committed:
Refactor: Judge protocol 삭제 -> Referee에서 모든 Ball과 Strike 판단

alstjr7437 committed:
refactor: Judgement 삭제 -> Referee에서 게임 결과 판단

alstjr7437 committed:
reafactor: 컴퓨터 숫자 변수명에서 상대방 숫자 변수명으로 변경

alstjr7437 committed:
refactor: restart 변수 삭제 guart를 통한 break로 변경

Redish03 commented on BaseBall/BaseBall/view/InputView.swift:
```diff
@@ -0,0 +1,20 @@
+class InputView {
+    func inputNumber() -> [Int]{
+        print("숫자를 입력해주세요 : ")
+        return readLine()!.compactMap { Int(String($0))}
```
`$0` 부분은 뭘 뜻하는 건가요?

alstjr7437 replied:
일단 기본적인 설명을 드리면!!

Swift에서 readLine()을 하게 되면 기본적으로 Opptional<String>으로 받아옵니다 
> Opptional -> nil이 가능한 변수

하지만 reaLine()!을 통해서 강제로 옵셔널을 벗겨서 nil이 불가능하게 만들었습니다

그리고 String을 CompactMap 함수를 사용하여 하나씩 짤라서 Int로 바꿔주는 부분입니다!!
그래서 질문으로 다시 돌아오면 $0은 각 문자열의 문자를 의미합니다
예로 들면 "123"이 들어오게 되면 ["1", "2", "3"]으로 들어와 Int로 바꿔주는 부분입니다!

```swift
let hello = "123"
print(hello.compactMap{$0}) // ["1", "2", "3"]

// 그리고 필터 등 여러가지에서 쓰입니다.
let array = [0,1,2,3]
let filteredArray = array.filter { $0 % 2 == 0}
print(filteredArray)   // [0, 2]
```

Redish03 commented on BaseBall/BaseBall/Controller/BaseBallGameController.swift:
```diff
@@ -0,0 +1,70 @@
+            
+        } catch {
+            outputView.errorInputPrint()
+        }
```
요구사항에 맞게 indent 2로 수정해 주세요!

alstjr7437 replied:
c060d8d55cd59100a56f1ab643300d42cac4de08 사실 이 커밋에 있는 것처럼 처음에는 depth를 맞춰서 작성했는데 변수로 while을 돌리는 부분은 좋지 않은 것 같아서 고민을 많이하다가 guard문을 썼습니다.

그래서 일단 depth를 다시 줄이기 위해 두가지 방법이 있었는데 do while과 비슷한 repeat while이 있는데 do while은 뭔가 땡기지 않아서... 아래와 같이 작성했습니다.

```swift
do {
    try playOneGame(opponentNumbers: OpponentBall(numberGenerator: RandomNumberGenerator()).numbers)
    while try restart() {
        try playOneGame(opponentNumbers: OpponentBall(numberGenerator: RandomNumberGenerator()).numbers)
    }
    outputView.endGamePrint()
    
} catch {
    outputView.errorInputPrint()
}

// repeat while 문법
do {
    repeat {
        try playOneGame(opponentNumbers: OpponentBall(numberGenerator: RandomNumberGenerator()).numbers)
    } while try restart()
    
    outputView.endGamePrint()
    
} catch {
    outputView.errorInputPrint()
}
```

Redish03 commented on BaseBall/BaseBall/Controller/BaseBallGameController.swift:
```diff
@@ -0,0 +1,70 @@
+//
+//  Created by KimMinSeok on 4/8/24.
+//
+enum RestartError: Error {
+    case IllegalArgumentException
+}
+
```
따로 enum 으로 IllegalArgumentException을 관리하는 이유는 Swift가 IllegalArgumentException이 없어서인가요????

alstjr7437 replied:
swift에서 에러로 [invalidArgumentException](https://developer.apple.com/documentation/foundation/nsexceptionname/1415426-invalidargumentexception)라는 것도 있지만 요구사항에는 IllegalArgumentException으로 반환하라고 적혀 있어서 그렇게 작성한 것 같네요

Enum으로 선언한 이유는 swift에서는 에러 핸들링을 하려면 우선적으로 enum으로 Error프로토콜을 이용하여 오류 처리 유형을 선언해주고 써줘야 해서 그렇게 작성했습니다.

Redish03 commented on BaseBall/BaseBall/domain/NumberGenerator/RandomNumberGenerator.swift:
```diff
@@ -0,0 +1,17 @@
+    func getNumbers() -> [Int] {
+        nums.shuffle()
+        return Array(nums[0...2])
+    }
```
클린코드 적용해서 상수화 시키는 건 어떤가요!

alstjr7437 replied:
상수화 시키면서 0...2의 숫자 길이를 조금 보기 편하게 하려고 아래 커밋과 같이 바꿨습니다!!
e890627417e6db10dc4752c648b5ab63e1f3d98d

Redish03 reviewed (commented):
수고하셨습니다! swift 가 kotlin이랑 좀 다르다 보니 보는데 막 술술 읽히진 않더라구요,,ㅎㅎㅋㅋㅎ 

로직 구현 보고 질문이나 간단한 코멘트 남겨놨습니다!

alstjr7437 committed:
feat: 점수 데이터 저장 구조체 구현

alstjr7437 committed:
feat: 상대방과 유저 수를 위한 Ball 프로토콜 구현 및 읽기 쓰기 선언

get으로 읽기 전용 선언해주고
let을 사용해봤지만 그렇게 되면 UserBall 생성시 에러 발생
그래서 private(set) var 으로 설정

alstjr7437 committed:
refactor: Guard 구문을 빼주고 depth 줄이기

alstjr7437 committed:
refactor: 숫자 부분 상수화 진행

SeongHoonC commented on BaseBall/BaseBall/domain/Ball/UserBall.swift:
```diff
@@ -0,0 +1,36 @@
+import Foundation
+
+class UserBall : Ball{
+    private(set) var numbers : [Int] = []
```
hasDuplication 이라는 함수로 중복을 확인하는데 [Int] 가 아닌 Set<Int> 를 사용해보는건 어떨까요?

alstjr7437 replied:
@Redish03 @SeongHoonC
사실 0과 중복을 받는 부분을 굳이 함수로 만들어야할까 초반 설계에서 고민을 했었습니다!!
그런데 매번 미션의 고민 중 하나가 해당 미션의 요구사항을 만족해서 작성하는데
다른 기능들을 추가하면 더 재밌고 발전할 수 있지 않을까였습니다!

다른 어플들도 마찬가지로 핵심 기능이 알약 관리 어플이였는데 영양제도 추가 되듯이

<br><br>

그러한 부분에서 이부분도 530 같은 숫자가 들어오게 설계도 가능하게 만드려면 함수를 추가해서 하면 되지 않을까 생각을 했었습니다!

그리고 중복 체크도 533같은 숫자가 들어오면 야구게임의 정확한 룰이 기억안나서 생각은 안나지만 그러한 부분도 있으면 더 어렵지만 재밌을 것 같아서 구현을 했는데 

<br><br>

다른분들은 어떻게 생각하시는지 궁금합니다!!

<br><br>

일단 요구사항으로 만들면 이렇게 됩니다!
사실 코드의 길이 부분은 줄어들지만 
추후에 0이 가능한지 안가능한지 받을때는 다시 코드를 짜줘야하는 불편함을 생각했었습니다.
```swift
enum BallCountError: Error {
    case IllegalArgumentException
}

import Foundation

class UserBall : Ball{
    
    private(set) var numbers : Set<Int> = []
    init(numbers: Set<Int>) throws {
        self.numbers = numbers
        
        guard numbers.count == BALL_LENGTH, !numbers.contains(INCLUDE_ZERO)else {
             throw BallCountError.IllegalArgumentException
        }

    }
    
    private let BALL_LENGTH = 3
    private let INCLUDE_ZERO = 0
    
}
```

Redish03 replied:
오호....제 기억상으론 중복되는 숫자는 불가능하다고 알고 있긴 합니다..! 
흠...룰을 바꾸는 것도 재밌을 것 같긴 하네요 저는 😄  그러면 관련 조건 체크를 함수로 뗄 것 같습니다ㅋㅎ

SeongHoonC replied:
음 요구사항에 유연하게 대응하는 것은 필요합니다만 이 객체를 만드는 이유를 생각해보셔야겠군요.

추가로 0이 아니다로 판단하면 11, -29484, 938282737 같은 숫자도 들어올 수 있지 않을까요?

SeongHoonC replied:
numbers: Set( 11, -29484, 938282737)

alstjr7437 replied:
따로 11 이렇게 들어오면 readline할때는 잘라서 바꿔버려서 생각을 안했었는데
도메인에서 저렇게 들어가버려 테스트하는 경우는 생각을 안했군요
그냥 하나씩 잘라서 갯수를 세서 해야할지 고민이 많은데 어떤 방법이 좋을까요?

SeongHoonC commented on BaseBall/BaseBall/domain/NumberGenerator/RandomNumberGenerator.swift:
```diff
@@ -0,0 +1,23 @@
+
+class RandomNumberGenerator : NumberGenerator{
+    
+    lazy var nums = Array(MIN_RANDOM_NUM...MAX_RANDOM_NUM)
```
상수(let)로 변경하고 shuffled() 를 사용할 수도 있겠네요

alstjr7437 replied:
989d8ff3dacffa293be55c1a891d5e6cf9f0be8a 과 같이 바꿨습니다!!
따로 왜 바꾼지 생각하는 부분도 설명부분에 추가했습니다

SeongHoonC commented on BaseBall/BaseBall/domain/Referee.swift:
```diff
@@ -0,0 +1,25 @@
+
+class Referee{
+    
+    func gameResultCount(inputNumbers1: [Int], inputNumbers2: [Int]) -> Score{
+        return Score(ball: ballCount(inputNumbers1: inputNumbers1, inputNumbers2: inputNumbers2), strike: strikeCount(inputNumbers1: inputNumbers1, inputNumbers2: inputNumbers2))
+    }
+    
+    func ballCount(inputNumbers1: [Int], inputNumbers2: [Int]) -> Int {
+        return inputNumbers1.filter { num in
+            inputNumbers2.contains(num) && num != inputNumbers2[inputNumbers1.firstIndex(of: num)!]
+        }.count
+    }
+    
+    func strikeCount(inputNumbers1: [Int], inputNumbers2: [Int]) -> Int {
+        return inputNumbers1.enumerated().filter { $0.element == inputNumbers2[$0.offset] }.count
+    }
```
함수 이름이 전체적으로 명사형입니다! getScore() 라던가 refereeScore() 등이 적당해 보이네요.

SeongHoonC replied:
inputNumber1, 2 보다는 기준(baseNumbers), 대상(targetNumbers)으로 나누는게 명확한 의미 전달이 될 것 같아요.

SeongHoonC replied:
외부에서 사용하지 않는 함수는 private 으로 만드는게 어떤가요?

alstjr7437 replied:
58da6eac9e66f130a151d2ef4aa7a525e14fc6bc
private를 다시 생각해보니 GameControlle에서도 private 선언하지 않은 부분이 많아서 수정했습니다!!
그리고 따로 커밋 내용에는 적혀있지 않지만 inputNumber보다는 저도 base 등이 보기 편한 것 같아서 수정했습니다!

SeongHoonC commented on BaseBall/BaseBall/domain/Score.swift:
```diff
@@ -0,0 +1,18 @@
+
+import Foundation
+
+struct Score {
+    var strike : Int
+    var ball : Int
+    
+    init(ball: Int, strike: Int) {
+        self.strike = strike
+        self.ball = ball
+    }
+}
```
struct 의 활용 :)

SeongHoonC replied:
strike 와 ball 이 var 이면 어떤 문제가 생길까요:?

alstjr7437 replied:
5d160b3d61859251c5711d9573c106d6a91a23e1
var로 하면 result로 얻어진 값이 추후에 변경이 가능해버려서 let으로 변경했습니다.

그리고 outputView에 굳이 Score라는 데이터 구조체를 알아야할까 해서 
아래와 같이 변경했습니다.
```swift
let resultScore = referee.getGameScore(baseNumbers: userNumbers, targetNumbers:opponentNumbers)
outputView.nowScorePrint(ball: resultScore.ball, strike: resultScore.strike)

func nowScorePrint(ball: Int, strike: Int){
    if ball == 0 && strike == 0 {
        print("낫씽")
        return
    }
    
    var output = ""
    if ball > 0 {
        output += "\(ball)볼 "
    }
    if strike > 0 {
        output += "\(strike )스트라이크"
    }
        
    print(output)
}
```

SeongHoonC commented on BaseBall/BaseBallTest/domain/OpponentBallTest.swift:
```diff
@@ -0,0 +1,19 @@
+final class OpponentBallTest: XCTestCase {
+
+
+    func test_상대방은_게임을_시작할때_랜덤한_숫자를_생성한다(){
+        let opponent = OpponentBall(numberGenerator: RandomNumberGenerator())
+        
+        XCTAssertNotEqual(opponent.numbers, [4,5,6])
+    }
```
Opponent 객체의 어떤 것을 테스트하기 위함이라고 보기 힘드네요.
항상 성공할 것으로 보이지도 않습니다.
랜덤 숫자를 생성하는 것이 필요한 테스트 혹은 가능한 테스트인가? 생각해봐야겠어요.

alstjr7437 replied:
922564c10d702f25f1c240f49867ddc037d94394
이부분에서 ball이라는 프로토콜을 만들다가 테스트의 사항이 애매해졌네요!!

상대가 번호를 잘 가지고 있는지는 테스트를 해야할 것 같아서
커밋 내용에도 적었듯이 위 3가지를 테스트 하는 사항으로 바꿨습니다!

SeongHoonC commented on BaseBall/BaseBallTest/domain/UserBallTest.swift:
```diff
@@ -0,0 +1,29 @@
+import XCTest
+
+final class UserBallTest: XCTestCase {
+
+    func test_숫자를_받아서_확인할_수_있다() throws{
+        let testBall = try UserBall(numbers: [3,2,4])
+        XCTAssertEqual(testBall.numbers, [3,2,4])
+    }
+  
+    func test_3자리의_수가_들어와야_한다() throws {
+        XCTAssertThrowsError(try UserBall(numbers: [1,2,3,4]))
+    }
+    
+    func test_숫자는_같은_수가_입력되면_안된다() throws {
+        XCTAssertThrowsError(try UserBall(numbers: [3,3,1]))
+    }
+    
+    func test_숫자는_0은_불가능하다() throws {
+        XCTAssertThrowsError(try UserBall(numbers: [3,3,0]))
+    }
+
```
예외 케이스 테스트 굿:)

SeongHoonC commented on BaseBall/BaseBallTest/domain/RefereeTest.swift:
```diff
@@ -0,0 +1,36 @@
+final class RefereeTest: XCTestCase {
+
+    func test_Referee는_볼을_출력한다(){
+        let number1 = [1,2,3]
+        let number2 = [3,4,5]
+        
+        let Referee = Referee()
+        XCTAssertEqual(Referee.ballCount(inputNumbers1: number1, inputNumbers2: number2), 1)
```
given, when, then 을 사용하면 가독성을 증가시킬 수 있습니다:)
[참고](https://www.swiftbysundell.com/articles/making-swift-tests-easier-to-debug/)

```suggestion
        // given
        let number1 = [1,2,3] 
        let number2 = [3,4,5]
        let referee = Referee()

        // when 
        let actual = referee.ballCount(inputNumbers1: number1, inputNumbers2: number2)
        
        // then
        XCTAssertEqual(actual, 1)
```

alstjr7437 replied:
3b6556425430addf29b138e95ec977c8262fd165 
given-when-then을 배웠는데 따로 TDD로 하면서 다 지우고 시작을 해서 까먹고 안해버렸네요!!

이제부터는 given-when-then을 모든 테스트코드에 적고 시작해야겠네요!

그리고 아래는 어떻게 given-when-then을 진행하면 좋을까요?
```swift
    func test_3자리의_수가_들어와야_한다() throws {
        XCTAssertThrowsError(try UserBall(numbers: [1,2,3,4]))
    }
    
    func test_숫자는_같은_수가_입력되면_안된다() throws {
        XCTAssertThrowsError(try UserBall(numbers: [3,3,1]))
    }
    
    func test_숫자는_0은_불가능하다() throws {
        XCTAssertThrowsError(try UserBall(numbers: [3,2,0]))
    }
```

Redish03 replied:
numbers가 given 이고 when은 비워야 하나...then에 XCTAssertThrowsError(try UserBall(numbers)) 일 것 같습니다..!

alstjr7437 replied:
```swift
// Given
let number1 = [1,2,3,4]

// Then 
XCTAssertThrowsError(try UserBall(numbers: number1))
```
성우님 말처럼 위와 같이 될 것 같네요

ad4ace890fa9a91255321f62b4da0169c76a2094

SeongHoonC replied:
given when then 은 가독성을 향상시키기 위함입니다.

질문의 예시로 보여준 코드는 간단하고 단계도 따로 없어 가독성이 좋다고 판단됩니다. 이런 경우에 굳이 모든 단계를 지키려고 분리하지않아도 됩니다!

SeongHoonC reviewed (changes requested):
리뷰 반영 확인했습니다! 
조금 더 고민했으면 좋겠다고 생각해서 리뷰 남겼습니다ㅎㅎ

반영 후 재요청해주세요!

alstjr7437 committed:
refactor: 랜덤 넘버 생성시 let으로 변경 후 shuffled로 변경

var -> let으로 변경하여서 중간에 값이 바뀌는 부분을 방지함

alstjr7437 committed:
refactor: 내부 함수 private 선언 및 함수명 동사형 수정

내부에서 돌아가는 함수는 public으로 둘 필요는 없으므로 private로 변경
- 게임 돌아가는 함수, 라운드 돌아가는 함수, 재시작 함수
- 스트라이크, 볼 판단 함수

함수명 동사형 수정
getBallCount 등 동사형으로 수정

alstjr7437 committed:
refactor: score 구조체 변경, outputView 매개변수 수정

따로 outputView가 Score라는 구조체를 알아야할까 생각이 들어서
매개변수를 넘겨줄때 그냥 ball 갯수와 Strike갯수를 바로 넘겨주는 식으로 변경했습니다.

alstjr7437 committed:
test: 상대방 테스트 사항 변경

상대방이 숫자를 가지는지
상대방이 0아 아닌 숫자를 가지는지
상대방이 중복이 되지 않는 숫자를 가지는지

3가지를 테스트 하도록 변경했습니다.

alstjr7437 committed:
test: given-when-then 패턴에 맞게 수정

given - 테스트 준비
when - 테스트 진행
then - 테스트 결과 검증

alstjr7437 committed:
test: 전체적인 test 수정(Given-When-Then)

alstjr7437 committed:
test: 에러 반환 함수 Given, Then 수정

SeongHoonC commented on BaseBall/BaseBall/Controller/BaseBallGameController.swift:
질문 2에 대한 답변입니다.
Controller 커밋 단위를 나누기는 좀 어렵긴하죠. 

함수가 단일 책임을 갖도록 잘 분리했다면 함수의 이름 짖기도 쉬워지고 커밋나누기도 편해질 것이라 예상됩니다.

SeongHoonC commented on BaseBall/BaseBall/Controller/BaseBallGameController.swift:
```diff
@@ -0,0 +1,69 @@
+        }
+    }
+    
+    private func playOneGame(opponentNumbers: [Int]) throws{
+        do {
+            var userNumbers = try UserBall(numbers: inputView.inputNumber()).numbers
+            while !userNumbers.elementsEqual(opponentNumbers){
+                playOneRound(userNumbers: userNumbers, opponentNumbers: opponentNumbers)
+                userNumbers = try UserBall(numbers: inputView.inputNumber()).numbers
+            }
+        } catch {
+            throw error
+        }
+        
+        outputView.resultGamePrint()
+    }
```
컴퓨터가 새로운 숫자를 만드는 것까지 게임의 책임으로 볼 수도 있겠네요~

alstjr7437 replied:
해당 부분은 잘 이해가 안가서 혹시 컴퓨터가 새로운 숫자를 만드는 것은 도메인에서 만들어져서 오는데 수정이 필요할까요???

SeongHoonC commented on BaseBall/BaseBall/Controller/BaseBallGameController.swift:
```diff
@@ -0,0 +1,69 @@
+    
+    func run() {
+        do {
+            try playOneGame(opponentNumbers: OpponentBall(numberGenerator: RandomNumberGenerator()).numbers)
+            while try restart() {
+                try playOneGame(opponentNumbers: OpponentBall(numberGenerator: RandomNumberGenerator()).numbers)
+            }
+            outputView.endGamePrint()
```
질문 3의 답변: 함수 순서는 보통 호출 순서와 동일하게 가면 됩니다:)

alstjr7437 replied:
감사합니다!! 위와 같이 
게임시작 -> 한판 시작 -> 한 라운드(유저가 숫자 입력) -> 재시작 여부 
순으로 순서를 정리해뒀습니다

SeongHoonC reviewed (changes requested):
리뷰반영 및 스스로 공부하는 모습까지 너무 멋있네요..
이전에 답변하지 못한 질문과 추가 리뷰까지 남겼습니다!
재요청해주세요~

SeongHoonC commented:
이것도 한번 확인해주세요~
https://github.com/wowap-course/kotlin-baseball/commit/922564c10d702f25f1c240f49867ddc037d94394#r141008952

alstjr7437 committed:
feat: 유저 숫자의 조건을 1부터 9까지로 바꿈

[1233, 2, 5]
고치기 전에는 위와 같은 숫자가 가능했음

alstjr7437 commented:
@SeongHoonC 
> 음 요구사항에 유연하게 대응하는 것은 필요합니다만 이 객체를 만드는 이유를 생각해보셔야겠군요.
> 
> 추가로 0이 아니다로 판단하면 11, -29484, 938282737 같은 숫자도 들어올 수 있지 않을까요?

해당 부분에서 5407e768e4912e6349614a2a59fef194b3e95ff8 의 커밋을 통해 0이 포함된 숫자인지 하지 않고
isSubset을 통해 1~9로 포함되어있는지 확인하고 테스트하게 바꿨습니다!!!
